### PR TITLE
New gadget: bee

### PR DESCRIPTION
**Pull request by alban**
# New gadget: bee

Based on https://github.com/solo-io/bumblebee

```
$ go run -exec sudo ./cmd/local-gadget/... trace bee --image="albantest.azurecr.io/bee-tcpconnect:dev2" -r docker
$ go run -exec sudo ./cmd/local-gadget/... trace bee --image="ghcr.io/solo-io/bumblebee/tcpconnect:0.0.11" -r docker
CONTAINER    PAYLOAD
             {"daddr":"192.168.49.2","saddr":"192.168.49.2"}
             {"daddr":"127.0.0.1","saddr":"127.0.0.1"}
             {"daddr":"192.168.49.2","saddr":"192.168.49.2"}
```

## How to use

* Clone https://github.com/solo-io/bumblebee
* Push your ebpf program in a OCI registry
```
$ bee build \
    examples/tcpconnect/tcpconnect.c \
    albantest.azurecr.io/bee-tcpconnect:dev \
    -i ghcr.io/solo-io/bumblebee/builder:0.0.9
$ bee push albantest.azurecr.io/bee-tcpconnect:dev
```

* Use `local-gadget trace bee --image ...`

## Testing done

Tested with the [tcpconnect](https://github.com/solo-io/bumblebee/blob/main/examples/tcpconnect/tcpconnect.c) ebpf program above.

## TODO

- [ ] container enrichment in userspace (pass mntnsid from ebpf to userspace)
- [ ] container enrichment in kernel (provide a "containers" ebpf map to the ebpf program updated with container-collection)
- [ ] container filtering
- [ ] generate columns dynamically
- [ ] support in kubectl-gadget (in addition to local-gadget)
- [ ] support ebpf programs other than kprobes/kretprobes
- [ ] support ebpf maps other than ring buffer
- [ ] upstream cilium/ebpf patches
- [ ] upstream bumblebee patches

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -177,4 +177,4 @@ replace (
 	github.com/gogo/protobuf v1.3.1 => github.com/gogo/protobuf v1.3.2
 )
 
-replace github.com/cilium/ebpf => github.com/inspektor-gadget/ebpf v0.0.0-20221115161606-78aefc9a99fd
+replace github.com/cilium/ebpf => github.com/inspektor-gadget/ebpf v0.0.0-20221124154204-19876eff65b2
```
[the remark this answers is not in the document]
Comment (alban) [reply go.mod:180]: Correct. Their branch is:
https://github.com/solo-io/cilium-ebpf/commits/ebpf-ext

Specifically this PR of Inspektor Gadget needs this patch:
https://github.com/inspektor-gadget/ebpf/commit/19876eff65b2f70a2b33f94a0e89d35aa326f45d
I pushed it on branch ig_bee.

**Comment by mauriciovasquezbernal** [comment]
I don't have specific comments about the implementation so far. 

I'm wondering about the TODO items:

> container enrichment in userspace (pass mntnsid from ebpf to userspace)

Do we want to make this automatically or the creator of the eBPF program should send it to us? 

> container filtering

Same as above. Do we want to implement this for the user or is it the user's responsibility to implement the filtering logic?

**Comment by alban** [comment]
> I'm wondering about the TODO items:
> 
> > container enrichment in userspace (pass mntnsid from ebpf to userspace)
> 
> Do we want to make this automatically or the creator of the eBPF program should send it to us? 
> 
> > container filtering
> 
> Same as above. Do we want to implement this for the user or is it the user's responsibility to implement the filtering logic? 

I implemented bpf extensions (bpf functions replaced dynamically) to provide additional bpf helper functions to enrich the traces with metadata, see example in #1193.

We could use similar logic for container filtering in bpf without requiring authors of bpf modules to have IG logic in their code.

They could use something like:
```
if (gadget_filter_container(ctx))
  return 0;
```
The helper `gadget_filter_container` can be written in a bpf extension. If IG changes the implementation of that extension, authors of external gadgets would not need to recompile their bpf code.

Similarly we can have:
```
event.container_cookie = gadget_get_container_cookie(ctx)
```
That would get the mntnsid. And if IG switches from using the mntnsid to cgroup-v2-id (e.g. for supporting systemd units in systemd-gadget), then authors of external gadgets would not need to recompile their bpf code.

**Comment by mauriciovasquezbernal** [comment]
@alban should we close in favor of https://github.com/inspektor-gadget/inspektor-gadget/pull/1527?

**Pull request closed without merging**